### ISSUE TITLE
Bug: splunk-operator subscription stuck UpgradePending

### ISSUE THREAD

**Issue by mrhillsman** [opdev/opcap#101]
When running the audit for OpenShift 4.8 the WaitForCsvOnNamespace fails:

```sh
Melvins-MacBook-Pro:opcap mrhillsman$ build/opcap check
I0722 01:08:44.870356   33816 request.go:601] Waited for 1.049241572s due to client-side throttling, not priority and fairness, request: GET:https://api.4-8-audit-install-config.opdev.io:6443/apis/image.openshift.io/v1?timeout=32s
check called
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x380 pc=0x2134739]

goroutine 1 [running]:
opcap/internal/operator.operatorClient.WaitForCsvOnNamespace({{0xc00021fce0?, 0x27b57b0?}}, {0xc001fa21f8, 0x15})
        /Users/mrhillsman/Development/goland/opcap/internal/operator/csv.go:57 +0x2f9
opcap/internal/capability.(*capAudit).OperatorInstall(0xc00164d2c0)
        /Users/mrhillsman/Development/goland/opcap/internal/capability/operator_install.go:42 +0x442
reflect.Value.call({0x2312f00?, 0xc00164d2c0?, 0x1?}, {0x250e7d4, 0x4}, {0x0, 0x0, 0x27c08e8?})
        /usr/local/opt/go/libexec/src/reflect/value.go:556 +0x845
reflect.Value.Call({0x2312f00?, 0xc00164d2c0?, 0xc001f7aed0?}, {0x0, 0x0, 0x0})
        /usr/local/opt/go/libexec/src/reflect/value.go:339 +0xbf
opcap/internal/capability.(*capAuditor).RunAudits(0xc002c7bd50)
        /Users/mrhillsman/Development/goland/opcap/internal/capability/auditor.go:78 +0x1f5
opcap/cmd.glob..func2(0x33e7ac0?, {0x250ecd0?, 0x0?, 0x0?})
        /Users/mrhillsman/Development/goland/opcap/cmd/check.go:62 +0xef
github.com/spf13/cobra.(*Command).execute(0x33e7ac0, {0x3433038, 0x0, 0x0})
        /Users/mrhillsman/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:860 +0x663
github.com/spf13/cobra.(*Command).ExecuteC(0x33e7d40)
        /Users/mrhillsman/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/mrhillsman/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
opcap/cmd.Execute()
        /Users/mrhillsman/Development/goland/opcap/cmd/root.go:36 +0x25
main.main()
        /Users/mrhillsman/Development/goland/opcap/main.go:10 +0x17
Melvins-MacBook-Pro:opcap mrhillsman$
```

![splunk](https://user-images.githubusercontent.com/3989695/180446053-659dd3f4-76dc-4546-9fdf-0d9b8b043a49.png)

Further investigation needs to be made to determine if we can reasonably handle such an error and continue audit.

**Comment by mrhillsman** [comment]
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: InstallPlan
metadata:
  generateName: install-
  resourceVersion: '241335'
  name: install-wkv44
  uid: 78519595-4584-46a7-873f-56994ae32e74
  creationTimestamp: '2022-07-22T13:33:09Z'
  generation: 1
  managedFields:
    - apiVersion: operators.coreos.com/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        'f:metadata':
          'f:labels':
            .: {}
            'f:operators.coreos.com/splunk-operator.opcap-splunk-operator': {}
      manager: olm
      operation: Update
      time: '2022-07-22T13:33:09Z'
    - apiVersion: operators.coreos.com/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        'f:metadata':
          'f:generateName': {}
          'f:ownerReferences':
            .: {}
            'k:{"uid":"3137a1a3-cd67-41d6-b52e-1a12933a8911"}':
              .: {}
              'f:apiVersion': {}
              'f:blockOwnerDeletion': {}
              'f:controller': {}
              'f:kind': {}
              'f:name': {}
              'f:uid': {}
        'f:spec':
          .: {}
          'f:approval': {}
          'f:approved': {}
          'f:clusterServiceVersionNames': {}
          'f:generation': {}
        'f:status':
          .: {}
          'f:bundleLookups': {}
          'f:catalogSources': {}
          'f:conditions': {}
          'f:phase': {}
      manager: catalog
      operation: Update
      time: '2022-07-22T13:33:10Z'
  namespace: opcap-splunk-operator
  ownerReferences:
    - apiVersion: operators.coreos.com/v1alpha1
      blockOwnerDeletion: false
      controller: false
      kind: Subscription
      name: splunk-operator
      uid: 3137a1a3-cd67-41d6-b52e-1a12933a8911
  labels:
    operators.coreos.com/splunk-operator.opcap-splunk-operator: ''
spec:
  approval: Automatic
  approved: true
  clusterServiceVersionNames:
    - splunk-operator.v1.1.0
  generation: 1
status:
  bundleLookups:
    - catalogSourceRef:
        name: certified-operators
        namespace: openshift-marketplace
      conditions:
        - message: bundle contents have not yet been persisted to installplan status
          reason: BundleNotUnpacked
          status: 'True'
          type: BundleLookupNotPersisted
        - message: unpack job not yet started
          reason: JobNotStarted
          status: 'True'
          type: BundleLookupPending
        - lastTransitionTime: '2022-07-22T13:33:09Z'
          message: Job has reached the specified backoff limit
          reason: BackoffLimitExceeded
          status: 'True'
          type: BundleLookupFailed
      identifier: splunk-operator.v1.1.0
      path: >-
        registry.connect.redhat.com/61b0dd9a4841f1221d73cdba/splunk-operator@sha256:72a0d007e387c1b8c6506ef6500f59014100c0f0fb53c6c710ebea917dd89da0
      properties: >-
        {"properties":[{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"ClusterMaster","version":"v1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"ClusterMaster","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"ClusterMaster","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"ClusterMaster","version":"v1beta2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"ClusterMaster","version":"v1beta3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"ClusterMaster","version":"v2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"ClusterMaster","version":"v3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"IndexerCluster","version":"v1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"IndexerCluster","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"IndexerCluster","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"IndexerCluster","version":"v1beta2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"IndexerCluster","version":"v1beta3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"IndexerCluster","version":"v2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"IndexerCluster","version":"v3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"LicenseMaster","version":"v1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"LicenseMaster","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"LicenseMaster","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"LicenseMaster","version":"v1beta2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"LicenseMaster","version":"v1beta3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"LicenseMaster","version":"v2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"LicenseMaster","version":"v3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"MonitoringConsole","version":"v3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"SearchHeadCluster","version":"v1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"SearchHeadCluster","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"SearchHeadCluster","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"SearchHeadCluster","version":"v1beta2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"SearchHeadCluster","version":"v1beta3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"SearchHeadCluster","version":"v2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"SearchHeadCluster","version":"v3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"Standalone","version":"v1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"Standalone","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"Standalone","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"Standalone","version":"v1beta2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"Standalone","version":"v1beta3"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"Standalone","version":"v2"}},{"type":"olm.gvk","value":{"group":"enterprise.splunk.com","kind":"Standalone","version":"v3"}},{"type":"olm.package","value":{"packageName":"splunk-operator","version":"1.1.0"}}]}
      replaces: ''
  catalogSources: []
  conditions:
    - lastTransitionTime: '2022-07-22T13:33:10Z'
      lastUpdateTime: '2022-07-22T13:33:10Z'
      message: >-
        Bundle unpacking failed. Reason: BackoffLimitExceeded, and Message: Job
        has reached the specified backoff limit
      reason: InstallCheckFailed
      status: 'False'
      type: Installed
  phase: Failed
```

**Comment by acmenezes** [comment]
Hi @mrhillsman , it seems that there may be a situation where the watcher.ResultChan sends a EOF as if the stop() or close() methods were called by the object without sending any events. That would force the code to jump the loop below and the string conversion on the return code will break with a nil pointer dereference or an invalid memory address:
```
	var csv *operatorv1alpha1.ClusterServiceVersion
	var ok bool

	for event := range watcher.ResultChan() {
		csv, ok = event.Object.(*operatorv1alpha1.ClusterServiceVersion)
		if !ok {
			return "", fmt.Errorf("received unexpected object type from watch: object-type %T", event.Object)
		}
		if csv.Status.Phase == operatorv1alpha1.CSVPhaseSucceeded ||
			csv.Status.Phase == operatorv1alpha1.CSVPhaseFailed {

			break
		}

	}

	return string(csv.Status.Phase), nil    // <----- this is line 57 where the error happens

```
I wonder if that's true and what would cause a watcher to send zero events and close the ResultChan prematurely. Specially if fails only with a single operator. I'll try to do further investigation on it. Is it always reproducible on your side? Are you seeing this for every run you try?

**Comment by acmenezes** [comment]
Just adding to the comment above. It seems that there are conditions where an error may actually occur and it's described on the package's documentation. Here is an excerpt from that:

> ResultChan returns a chan which will receive all the events. If an error occurs or Stop() is called, the implementation will close this channel and release any resources used by the watch.

I'll take this one and implement error handling there.

**Comment by mrhillsman** [comment]
Fixed by #112
